metricbeat/module/vsphere/resourcepool/resourcepool.go

```diff
@@ -0,0 +1,153 @@
+// MetricSet has been created, Fetch will be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("vsphere", "resourcepool", New,
+		mb.WithHostParser(vsphere.HostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet type defines all fields of the MetricSet.
+type ResourcePoolMetricSet struct {
+	*vsphere.MetricSet
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := vsphere.NewMetricSet(base)
+	if err != nil {
+		return nil, err
+	}
+	return &ResourcePoolMetricSet{ms}, nil
+}
+
+// Structure to hold performance metrics values
+type metricData struct {
+	assetNames assetNames
+}
+
+type assetNames struct {
+	outputVmNames []string
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *ResourcePoolMetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	client, err := govmomi.NewClient(ctx, m.HostURL, m.Insecure)
+	if err != nil {
+		return fmt.Errorf("error in NewClient: %w", err)
+	}
+
+	defer func() {
+		if err := client.Logout(ctx); err != nil {
+			m.Logger().Errorf("error trying to log out from vSphere: %w", err)
+		}
+	}()
+
+	c := client.Client
+
+	// Create a view of ResourcePool objects.
+	mgr := view.NewManager(c)
+
+	v, err := mgr.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"ResourcePool"}, true)
+	if err != nil {
+		return fmt.Errorf("error in CreateContainerView: %w", err)
+	}
+
+	defer func() {
+		if err := v.Destroy(ctx); err != nil {
+			m.Logger().Errorf("error trying to destroy view from vSphere: %w", err)
+		}
+	}()
+
+	// Retrieve property for all ResourcePools.
+	var rps []mo.ResourcePool
+	err = v.Retrieve(ctx, []string{"ResourcePool"}, []string{"name", "overallStatus", "vm", "summary"}, &rps)
+	if err != nil {
+		return fmt.Errorf("error in Retrieve: %w", err)
+	}
+
+	pc := property.DefaultCollector(c)
+	for i := range rps {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			assetNames, err := getAssetNames(ctx, pc, &rps[i])
+			if err != nil {
+				m.Logger().Errorf("Failed to retrieve object from resource pool %s: %w", rps[i].Name, err)
+			}
+
+			reporter.Event(mb.Event{
+				MetricSetFields: m.mapEvent(rps[i], &metricData{assetNames: assetNames}),
+			})
+		}
+	}
+
+	return nil
+}
+
+func getAssetNames(ctx context.Context, pc *property.Collector, rp *mo.ResourcePool) (assetNames, error) {
+	referenceList := append([]types.ManagedObjectReference{}, rp.Vm...)
+
+	var objects []mo.ManagedEntity
+	if len(referenceList) > 0 {
+		if err := pc.Retrieve(ctx, referenceList, []string{"name"}, &objects); err != nil {
```

Comment:
You could use rp.Vm directly
```suggestion
		if err := pc.Retrieve(ctx, rp.Vm, []string{"name"}, &objects); err != nil {
```

Reply:
I see this has been addressed, marking as resolved.